### PR TITLE
Synchronization Validation Phase 1 -- alpha 1

### PR DESCRIPTION
**Pull request by jzulauf-lunarg**
First set of synchronization validation features and unit level test:

Feature set: 
* Hazard detection for memory usage for commands within the *same* command buffer
* Synchronization operations vkCmdPipelineBarrier and renderpass/subpass barriers

Enabled by  placeholder feature "VK_VALIDATION_FEATURE_ENABLE_SYNCHRONIZATION_VALIDATION"  -- extension to VK_EXT_validation_features to follow merge to master.

Testing (for correctness, no known crashes) against applications and CTS ongoing.

Design doc added to docs/

**TODO/KNOWN LIMITATION:** Does not include implementation of multi-view renderpass support. 
**TODO/KNOWN LIMITATION:** Does not include vkCmd(Set|Wait)Event support. 
**TODO/KNOWN LIMITATION:** Host set event not supported
**TODO/KNOWN LIMITATION:** ExecuteCommands and QueueSubmit not supported in phase 1
**TODO/KNOWN LIMITATION:** Memory access checks not suppressed for  VK_CULL_MODE_FRONT_AND_BACK
**TODO/KNOWN LIMITATION:** First phase does not include component granularity access tracking.  
**TODO/KNOWN LIMITATION:** Host synchronization not supported

### CODE REVIEW

File: docs/synchronization.md [hunk 1]
```diff
@@ -0,0 +1,1332 @@
...
+   <td>Occurs when unsynchronized subpasses/queues perform read and write operations on the same set of memory locations
+   </td>
+  </tr>
+</table>
+
+
+
+### Algorithmic Overview {#algorithmic-overview}
+
+In order to validate synchronization the effect of action and synchronization commands must be tracked.  The full validation can *only* be done at queue submission time, when the full context of the command batches is known.  The full validation can *only* be done at queue submission time, when the full context of the command batches is known. However, partial validation of hazards knowable within the context of a single recorded command buffer can be done at record time at. This partial validation includes record time detection of internally present hazards during sub-pass and can be extended to include secondary command buffer execution. To support this usage, the validation and update components must be designed to operate at record, secondary command buffer execute record, and queue submit time.
```
Comment (filnet) [inline docs/synchronization.md:76]: The sentence that starts with "The full validation can ..."  is duplicated.

File: docs/synchronization.md [hunk 2]
```diff
@@ -0,0 +1,1332 @@
...
+Authors: Tobin Ehlis, John Zulauf
+
+Toward a unified model of Vulkan Synchronization
+
+
+[TOC]
+
+
+
+### Overview {#overview}
```
[the remark this answers is not in the document]
Comment (jzulauf-lunarg) [reply docs/synchronization.md:15]: This is auto generated from google docs which is still the controlling copy...

File: docs/synchronization.md [hunk 3]
```diff
@@ -0,0 +1,1332 @@
...
... access and synchronization operations for a range of memory address against a current access or operation against that same set. The _state_ (Resource Access State) tracks the most recent read, write, and synchronization operations for a given range of memory addresses. These are captured in _contexts _(Access Context) which reflect the Queue, Command Buffer, or Subpass instance in which the memory accesses were made. This first approach is sufficient for single command buffer validation. Note that only the most recent state (relative to *write* operations)  is retained. This reflects an assumption that given a sequence of writes _W(j)_, any “transitive hazard” write _W(i-2)_ conflicts with write _W(i)_, would also result in either a hazard being detected between _W(i-2) _and _W(i-1)_ or between _W(i-1)_ and _W(i)_ s.t. for purposes of detection, only the most recent state is required.  Additionally, the “dependency chain” logic can be represented by a cumulative state reflecting the impact of all synchronization operations on the most recent read/write state. Between the _state _and _context_, enough information is present to do record time validation of single command buffers. \
```
[the remark this answers is not in the document]
Comment (jzulauf-lunarg) [reply docs/synchronization.md:79]: @jeremy-lunarg -- state and context storage is defined below. are you looking for something else/more/different?

File: docs/synchronization.md (hunk 1 above)
Comment (jzulauf-lunarg) [reply docs/synchronization.md]: removed

**Comment by jzulauf-lunarg** [comment]
This should be considered pre-alpha.  Testing against applications and CTS is ongoing.

**Pull request merged**